boa/src/builtins/json/mod.rs

```diff
@@ -72,59 +72,136 @@ impl Json {
+        // 9. Let unfiltered be completion.[[Value]].
+        // 10. Assert: unfiltered is either a String, Number, Boolean, Null, or an Object that is defined by either an ArrayLiteral or an ObjectLiteral.
+        let unfiltered = context.eval(script_string.as_bytes())?;
+
+        match args.get(1).cloned().unwrap_or_default().as_object() {
+            // 11. If IsCallable(reviver) is true, then
+            Some(obj) if obj.is_callable() => {
+                // a. Let root be ! OrdinaryObjectCreate(%Object.prototype%).
+                let root = context.construct_object();
+
+                // b. Let rootName be the empty String.
+                // c. Perform ! CreateDataPropertyOrThrow(root, rootName, unfiltered).
+                root.create_data_property_or_throw("", unfiltered, context)
+                    .expect("CreateDataPropertyOrThrow should never throw here");
+
+                // d. Return ? InternalizeJSONProperty(root, rootName, reviver).
+                Self::internalize_json_property(root, "".into(), obj, context)
             }
-            Err(err) => context.throw_syntax_error(err.to_string()),
+            // 12. Else,
+            // a. Return unfiltered.
+            _ => Ok(unfiltered),
         }
     }
 
-    /// This is a translation of the [Polyfill implementation][polyfill]
+    /// `25.5.1.1 InternalizeJSONProperty ( holder, name, reviver )`
     ///
-    /// This function recursively walks the structure, passing each key-value pair to the reviver function
-    /// for possible transformation.
+    /// More information:
+    ///  - [ECMAScript reference][spec]
     ///
-    /// [polyfill]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse
-    fn walk(
-        reviver: &JsValue,
+    /// [spec]: https://tc39.es/ecma262/#sec-internalizejsonproperty
+    fn internalize_json_property(
+        holder: JsObject,
+        name: JsString,
+        reviver: JsObject,
         context: &mut Context,
-        holder: &mut JsValue,
-        key: &PropertyKey,
     ) -> JsResult<JsValue> {
-        let value = holder.get_field(key.clone(), context)?;
-
-        if let JsValue::Object(ref object) = value {
-            let keys: Vec<_> = object.borrow().properties().keys().collect();
+        // 1. Let val be ? Get(holder, name).
+        let val = holder.get(name.clone(), context)?;
+
+        // 2. If Type(val) is Object, then
+        if let Some(obj) = val.as_object() {
+            // a. Let isArray be ? IsArray(val).
+            // b. If isArray is true, then
+            if obj.is_array() {
+                // i. Let I be 0.
+                // ii. Let len be ? LengthOfArrayLike(val).
+                let mut i = 0;
+                let len = obj.length_of_array_like(context)? as i64;
+
+                // iii. Repeat, while I < len,
+                while i < len {
```

Comment:
```suggestion
                for i in 0..len {
```
Better to use a range instead of an additional variable.

Reply:
Done